### DISCUSSION TITLE
carelink login error

### DISCUSSION THREAD

**Discussion by seohyeonyang** [NightscoutFoundation/xDrip#3005] [Q&A]
response code : 504

It's so hard. Please help me.

**Comment by ranvens** [comment]
I'm getting the same 504 trying to Login to Au, with both patient and follower creds. Used xdrip years ago but recently changed to carelink and trying to find alternatives to the medtronic app due to constant outages.

**Comment by Navid200** [comment]
If you use the Carlink app, does it show readings?

**Reply by ranvens** [reply]
Yes -  it is currently showing data via carelink app.

**Reply by Navid200** [reply]
You can leave it active and use the companion app mode to get readings in xDrip:
https://navid200.github.io/xDrip/docs/Follow/CompanionApp.html

**Reply by illuzionist** [reply]
hi. I'm trying to configure Xdrip+ for local data transfer from Guardian CGM. but the data is not transferred to Xdrip+. 
1. I did the following: installed the data source "Companion Application" in Xdrip+
2. gave permission to view Xdrip + alerts. 
3. I left only two constant alerts with sugar data - Xdrip+ and Guardian CGM
4. stopped the sensor and started the sensor again. according to this instruction. https://xdrip.readthedocs.io/en/latest/install/companion/#allow-notification-access-greyed-out
but the data does not arrive in Xdrip+. I have attached a screenshot of the error notification. What am I doing wrong? 
I have already downloaded the new version and installed it on the phone, I have not imported the old settings, I have configured everything again, but it does not work. samsung a71 phone, android 13
![Screenshot_20230814_161250_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/581ba4bf-e2a5-49ec-a379-25ae17323548)

**Reply by ah7675** [reply]
@illuzionist To debug companion app, please add 'UiBasedCollector:v' to "less common settings" -> "extra loggoing settings" - > "extra tag". Waiting for 1-2 mins and then capture event log.

**Reply by illuzionist** [reply]
@ah7675 
here is a screenshot. a few days ago, I opened a discussion on this issue separately at this link. Thank you.
https://github.com/NightscoutFoundation/xDrip/discussions/3025
![Screenshot_20230817_162443_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/b6d5fe49-1807-4384-89e2-e331a8fe35c8)

**Comment by ah7675** [comment]
```
08-17 19:23:31.096 32209  5353 D OkHttp  : --> GET https://carelink.minimed.eu/patient/sso/login?country=tw&lang=en
08-17 19:23:31.097 32209  5353 D OkHttp  : Accept-Language: en;q=0.9, *;q=0.8
08-17 19:23:31.097 32209  5353 D OkHttp  : sec-ch-ua: "Chromium";v="112", "Google Chrome";v="112", "Not:A-Brand";v="99"
08-17 19:23:31.097 32209  5353 D OkHttp  : User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36
08-17 19:23:31.097 32209  5353 D OkHttp  : Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
08-17 19:23:31.097 32209  5353 D OkHttp  : Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
08-17 19:23:31.097 32209  5353 D OkHttp  : --> END GET
08-17 19:23:31.144 32209  5353 D OkHttp  : <-- 504 Gateway Time-out https://carelink.minimed.eu/patient/sso/login?country=tw&lang=en (47ms)
```


Simply change version tag in HTTP header from 112 to 115 fix the problem in my case.
But I'm not sure what's the root cause.
```
.addHeader("sec-ch-ua", "\"Chromium\";v=\"115\", \"Google Chrome\";v=\"115\", \"Not:A-Brand\";v=\"99\"")
.addHeader("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36");
```

**Reply by sarhatabaot** [reply]
In which file do I change this setting?
My dad is experiencing this issue as well.

Obv I have no issue compiling it but it seems to be related to a previous issue that was fixed in may.

**Reply by ah7675** [reply]
app/src/main/java/com/eveningoutpost/dexdrip/cgm/carelinkfollow/client/CareLinkClient.java

In fact, I'm inspired by that issue.

**Reply by sarhatabaot** [reply]
Perfect that worked for us :)

**Reply by illuzionist** [reply]
I also have the error "response code : 504"
did I understand correctly that you managed to fix it?
if so, when will it be possible to download the corrected version? thank you!

**Comment by Navid200** [comment]
@benceszasz

**Comment by ah7675** [comment]
Please review https://github.com/NightscoutFoundation/xDrip/pull/3029

**Comment by illuzionist** [comment]
I downloaded Nightly build 19th Aug 2023 and I have a 504 error gone.

**Comment by thanasistrisp** [comment]
Issue continues for me for last few days. I get login failed.

**Reply by sarhatabaot** [reply]
I'm experiencing this again as well

**Reply by thanasistrisp** [reply]
```
2023-09-29 19:58:49.565 28061-28225 CareLinkFollowDL        com.eveningoutpost.dexdrip           E  CareLink login error!  Response code: 200
2023-09-29 19:58:49.584 28061-28225 CareLinkFollowDL        com.eveningoutpost.dexdrip           E  Error message: NullPointerException:value == null
```

**Reply by Klinu** [reply]
My wife have same login failure now about two days. We are located in Finland.

**Reply by Liverpool4** [reply]
Yes, me too since about Wednesday - 5 days. Login error, with response code 200 (Spain).

There was a Medtronic (EU?) update to carelink iirc this week, does the http header need to be updated?

**Reply by cristiangomezm85** [reply]
Medtronic has added human verification to login on carelink ("I'm not a robot" captcha).

Since this update, we can't login from xdrip due this validation change. I've seen this new captcha appears now in all European servers.

We hope Medtronic rollback this change...

**Reply by aisaila** [reply]
Can you open a login widow to pass the captcha and then save the session?

**Reply by cristiangomezm85** [reply]
@aisaila I can login on web with captcha, but how can I save the session? Xdrip try to open a different login without captcha.

Is there a way to put logged session manually on xdrip?

**Reply by benceszasz** [reply]
I have released a first working prototype of my private CareLinkFollower for xdrip app for reCAPTCHA. If it is working well, I will create a new pull request for the official xdrip. But first there are few things to do, for example to use reCAPTCHA browser login only for servers outside US, since in the US the reCAPTCHA is not required (yet?).

**Reply by cristiangomezm85** [reply]
@benceszasz thanks for your dedication to solve this! I've seen captcha verification only appears on European servers (US not yet). If you need European user to make testing, contact me. Thanks!

**Reply by cristiangomezm85** [reply]
@benceszasz your version is working perfectly! After my first web login, +15hours reading constantly. Great job! Thanks!

**Reply by aisaila** [reply]
also installed and it works. GG

**Reply by aisaila** [reply]
looks like after 24h the token expires and I need to login again

**Reply by cristiangomezm85** [reply]
@aisaila not in my case. I've made web login 48h ago, and xdrip still working without login needed

**Reply by Liverpool4** [reply]
@cristiangomezm85 which variant of beta13 are you using? I installed variant4.apk and after an hour it didn't renew the session (Spain). 

I've just installed the xdrip-carelink.apk - the base version - I'll see how that goes.

**Reply by aisaila** [reply]
@cristiangomezm85 I use variant 4

**Reply by cristiangomezm85** [reply]
On app appears me the name of variant1. Someone send the file base.apk last Sunday in a diabetes WhatsApp group. I've installed this version, I've logged in using web login, and since then no more login required (I'm from Spain too)

**Reply by aisaila** [reply]
Ok I will try variant 1  then, thanks

**Reply by cristiangomezm85** [reply]
Is there a way on xdrip to screenshot some screen to share with you the exact name of the version?

**Reply by Liverpool4** [reply]
Interesting - I'm using the base from benceszasz's listing here on GitHub, it's been running perfectly for about 4 hours (I got an couldn't renew session error after an hour with variant4). 

And cristiangomez on variant1, and aisaila on variant4

**Reply by Liverpool4** [reply]
As you say, the variant appears in teh app name, so that's fine.

I was wondering what the differences were between the 5 apks (available via the dev's profile under the v0.1.13-beta) - and which is the most stable!

**Reply by benceszasz** [reply]
All variants are the same, the only difference is the app name / app id. This way multiple instances can be run on one phone to follow multiple persons or to install different version. For example I am running 5 variants (xdip, va1, ... va4) on one of my test phones and receiving data from CareLink for multiple accounts (patient, caregiver,...).

**Reply by aisaila** [reply]
I switched to varient 1 (does not matter if they are the same) and it looks like I sometimes get an "Error renewing token!" and that is the time that i need to r reconnect. I think it happens if i have a bad internet connection.

**Reply by thanasistrisp** [reply]
Me too I got token expired after two days about (Greece).

**Reply by benceszasz** [reply]
CareLink Follower with reCAPTCHA using browser login has been merged into the official xDrip version yesterday (08/10/2023). Latest official xDrip release already contains this new version of CareLink Follower with reCAPTCHA using browser login:
https://github.com/NightscoutFoundation/xDrip/releases/tag/2023.10.08
Please try to use a dedicated CareLink Care Partner account for xDrip to avoid closing your xDrip session by logging in with the same account from either the CareLink Personal website or a Medtronic offical app.

**Comment by Liverpool4** [comment]
I have been using the app since this time Tuesday, I'm located in Spain.

Apart from an initial glitch when the first token renewal didn't go according to plan, the app has run non-stop without issues for 3 days (72 hours)

Also, as a result I've been able to view it on my Garmin watch via the same Garmin widget as before the login change, as well as the watch picking up the blood glucose datafield on workouts as previously.

Nice work benceszasz!
